### PR TITLE
Update stateful_set_utils.go

### PR DESCRIPTION
**Pull request by ivanszl**
Signed-off-by: ivan.lan [ivan.lin.1985@gmail.com](mailto:ivan.lin.1985@gmail.com)

### Ⅰ. Describe what this PR does
设置sts的spec.podManagementPolicy = Parallel
spec.scaleStrategy.maxUnavailable = 100%
由于修改sts的模版，导致Pod启动不了，状态一直处于pending，后续重新修改会卡住
重新修改后会作用于所有的Pod，可以快速修复业务服务

### CODE REVIEW

**Comment by zmberg** [comment]
@ivanszl tks, and please *git commit --amend --no-edit --signoff*, and git push -f again.

**Comment by ivanszl** [comment]
> @ivanszl Can you describe the process in detail? I did not reproduce your problem. In addition, I re-reviewed the code and feel that the current val <= 0 judgment logic should meet expectations.

hi, this is a example
```yaml
kind: StatefulSet
apiVersion: apps.kruise.io/v1beta1
metadata:
  name: busybox-test
  namespace: default
  labels:
    app: busybox-test
spec:
  replicas: 1
  selector:
    matchLabels:
      app: busybox-test
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: busybox-test
    spec:
      containers:
        - name: busybox
          image: busybox:latest
          resources: {}
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          imagePullPolicy: IfNotPresent
      restartPolicy: Always
      terminationGracePeriodSeconds: 30
      dnsPolicy: ClusterFirst
      securityContext: {}
      schedulerName: default-scheduler
  podManagementPolicy: Parallel
  scaleStrategy:
    maxUnavailable: 100%
  updateStrategy:
    type: RollingUpdate
    rollingUpdate:
      partition: 0
      maxUnavailable: 100%
  revisionHistoryLimit: 10
```
```shell
$ kubectl get pods busybox-test-0
NAME             READY   STATUS             RESTARTS      AGE
busybox-test-0   0/1     CrashLoopBackOff   5 (95s ago)   4m45s
```
modify yaml and apply it
```yaml
kind: StatefulSet
apiVersion: apps.kruise.io/v1beta1
metadata:
  name: busybox-test
  namespace: default
  labels:
    app: busybox-test
spec:
  replicas: 1
  selector:
    matchLabels:
      app: busybox-test
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: busybox-test
        test-label: 'yes'
    spec:
      containers:
        - name: busybox
          image: busybox:latest
          resources: {}
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          imagePullPolicy: IfNotPresent
      restartPolicy: Always
      terminationGracePeriodSeconds: 30
      dnsPolicy: ClusterFirst
      securityContext: {}
      schedulerName: default-scheduler
  podManagementPolicy: Parallel
  scaleStrategy:
    maxUnavailable: 100%
  updateStrategy:
    type: RollingUpdate
    rollingUpdate:
      partition: 0
      maxUnavailable: 100%
  revisionHistoryLimit: 10
```
and then look the pod
```shell
kubectl describe pods busybox-test-0
Name:         busybox-test-0
Namespace:    default
Priority:     0
Node:         minikube/192.168.49.2
Start Time:   Wed, 30 Nov 2022 15:16:28 +0800
Labels:       app=busybox-test
              controller-revision-hash=busybox-test-f89f864cc
              lifecycle.apps.kruise.io/state=Normal
              statefulset.kubernetes.io/pod-name=busybox-test-0
Annotations:  lifecycle.apps.kruise.io/timestamp: 2022-11-30T07:16:28Z
Status:       Running
IP:           172.17.0.6
IPs:
  IP:           172.17.0.6
Controlled By:  StatefulSet/busybox-test
Containers:
  busybox:
    Container ID:   docker://a1227f31ef27e0d4fbf61aa29f585da912ce5dbe2fcc662a1e37a2ca3b4329b0
    Image:          busybox:latest
    Image ID:       docker-pullable://busybox@sha256:fcd85228d7a25feb59f101ac3a955d27c80df4ad824d65f5757a954831450185
    Port:           <none>
    Host Port:      <none>
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Wed, 30 Nov 2022 15:22:20 +0800
      Finished:     Wed, 30 Nov 2022 15:22:20 +0800
    Ready:          False
    Restart Count:  6
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-bml2r (ro)
Readiness Gates:
  Type             Status
  KruisePodReady   True
Conditions:
  Type              Status
  KruisePodReady    True
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  kube-api-access-bml2r:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason     Age                    From               Message
  ----     ------     ----                   ----               -------
  Normal   Scheduled  6m25s                  default-scheduler  Successfully assigned default/busybox-test-0 to minikube
  Normal   Pulling    6m24s                  kubelet            Pulling image "busybox:latest"
  Normal   Pulled     6m17s                  kubelet            Successfully pulled image "busybox:latest" in 6.646627307s
  Normal   Created    4m41s (x5 over 6m17s)  kubelet            Created container busybox
  Normal   Started    4m41s (x5 over 6m17s)  kubelet            Started container busybox
  Normal   Pulled     4m41s (x4 over 6m16s)  kubelet            Container image "busybox:latest" already present on machine
  Warning  BackOff    83s (x26 over 6m15s)   kubelet            Back-off restarting failed container
```
we find the pod was't modified

**Comment by zmberg** [comment]
@ivanszl Do you only want to modify labels in this release? Is it better to change the image or env?

**Comment by zmberg** [comment]
@ivanszl I don't think it can be modified directly, because it will destroy the semantics of [scaleStrategy](https://openkruise.io/docs/user-manuals/advancedstatefulset#scaling-with-rate-limiting).

The core of this problem is the corner case for 100% maxUnavailable, so i think it can be modified as follows:
```
if set.Spec.ScaleStrategy != nil && set.Spec.ScaleStrategy.MaxUnavailable != nil {
...
scaleMaxUnavailable = &maxUnavailable
              // add the code, to solve the 100% maxUnavailable corner case. 
    if int32(*scaleMaxUnavailable) == *set.Spec.Replicas {
	    scaleMaxUnavailable = utilpointer.IntPtr(math.MaxInt32)
    }
}
```

**Comment by ivanszl** [comment]
@zmberg If it was modified like you code, it can't cover same case. 
If you are set maxUnavailable < set.Spec.Replicas, you can find the updated pods not equal maxUnavailable, it's always equal maxUnavailable - 1.

**Comment by zmberg** [comment]
@ivanszl You can look at this document https://openkruise.io/docs/user-manuals/advancedstatefulset/#scaling-with-rate-limiting. Your modification method will break the semantics of the document

**Comment by ivanszl** [comment]
@zmberg Ok, I see.
It will break the semantics.

**Comment by zmberg** [comment]
@ivanszl We also discussed it internally, and felt that scaleStrategy should only affect Create Pod, not Update Pod, so i think there are three places that need to be modified, for example:
```
// if the set does not allow bursting, return immediately
if monotonic {
      return &status, nil
} else if decreaseAndCheckMaxUnavailable(scaleMaxUnavailable) {
     break
}
```

**Comment by ivanszl** [comment]
> @ivanszl We also discussed it internally, and felt that scaleStrategy should only affect Create Pod, not Update Pod, so i think there are three places that need to be modified, for example:
> 
> ```
> // if the set does not allow bursting, return immediately
> if monotonic {
>       return &status, nil
> } else if decreaseAndCheckMaxUnavailable(scaleMaxUnavailable) {
>      break
> }
> ```
@zmberg 
I think this places that need to modified may be better, modified as follow
```
// if we find no more Pods can be created, no more work can be done this round
if decreaseAndCheckMaxUnavailable(scaleMaxUnavailable) {
     break
}
....
// if the set does not allow bursting, return immediately
if monotonic {
    return &status, nil
}
```

**Comment by zmberg** [comment]
The monotonic parameter is still more preferred, and the scale Strategy is only considered when it is false.

**Comment by ivanszl** [comment]
@zmberg 
I find other call decreaseAndCheckMaxUnavailable function places, it was decrease and check before to do, but on the create block was decrease and check after created, so i think this modified may be preferred.

**Comment by zmberg** [comment]
> @zmberg I find other call decreaseAndCheckMaxUnavailable function places, it was decrease and check before to do, but in the create block was decrease and check after created, so i think this modified may be preferred.

@ivanszl Relevant places need to be modified. The above is just an example, not all of them are listed.

File: pkg/controller/statefulset/stateful_set_utils.go [hunk 1]
```diff
@@ -687,5 +687,5 @@ func decreaseAndCheckMaxUnavailable(maxUnavailable *int) bool {
 	}
 	val := *maxUnavailable - 1
 	*maxUnavailable = val
-	return val <= 0
+	return val < 0
```
Comment (zmberg) [inline pkg/controller/statefulset/stateful_set_utils.go:690]: There should be no further adjustments here, and the original writing method should be maintained.

File: pkg/controller/statefulset/stateful_set_control.go [hunk 2]
```diff
@@ -542,7 +547,7 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 				status.UpdatedReplicas++
 			}
 			// if the set does not allow bursting, return immediately
-			if monotonic || decreaseAndCheckMaxUnavailable(scaleMaxUnavailable) {
+			if monotonic {
 				return &status, nil
 			}
```
Comment (zmberg) [inline pkg/controller/statefulset/stateful_set_control.go:552]:
```
if monotonic {
    return &status, nil
} else if decreaseAndCheckMaxUnavailable(scaleMaxUnavailable) {
    break
}

if isTerminating(replicas[i]) && monotonic {
	klog.V(4).Infof(
		"StatefulSet %s/%s is waiting for Pod %s to Terminate",
		set.Namespace,
		set.Name,
		replicas[i].Name)
	return &status, nil
} else if isTerminating(replicas[i]) && decreaseAndCheckMaxUnavailable(scaleMaxUnavailable) {
     break
}

if monotonic || scaleMaxUnavailable != nil {
	isAvailable, waitTime := isRunningAndAvailable(replicas[i], minReadySeconds)
	if !isAvailable && monotonic {
		if waitTime > 0 {
			// make sure we check later
			durationStore.Push(getStatefulSetKey(set), waitTime)
			klog.V(4).Infof(
				"StatefulSet %s/%s needs to wait %s for the Pod %s to be Running and Available after being"+
					" Ready for %d seconds",
				set.Namespace,
				set.Name,
				waitTime,
				replicas[i].Name,
				minReadySeconds)
		} else {
			klog.V(4).Infof(
				"StatefulSet %s/%s is waiting for Pod %s to be Running and Ready",
				set.Namespace,
				set.Name,
				replicas[i].Name)
		}
		return &status, nil
	} else if !isAvailable && decreaseAndCheckMaxUnavailable(scaleMaxUnavailable) {
		break
	}
}
```

Comment (zmberg) [reply pkg/controller/statefulset/stateful_set_control.go:552]: If it is convenient for you, you can communicate on the zoom of the community meeting on 12.8.

File: pkg/controller/statefulset/stateful_set_control.go [hunk 3]
```diff
@@ -542,21 +542,25 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 				status.UpdatedReplicas++
 			}
 			// if the set does not allow bursting, return immediately
-			if monotonic || decreaseAndCheckMaxUnavailable(scaleMaxUnavailable) {
+			if monotonic {
 				return &status, nil
+			} else if decreaseAndCheckMaxUnavailable(scaleMaxUnavailable) {
+				break
```
Comment (zmberg) [inline pkg/controller/statefulset/stateful_set_control.go:548]: klog.V(4).Infof(
					"StatefulSet %s/%s Pod %s is Creating, and break pods scale",
					set.Namespace,
					set.Name,
					replicas[i].Name)

File: pkg/controller/statefulset/stateful_set_control.go [hunk 4]
```diff
@@ -542,21 +542,25 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
...
-		if isTerminating(replicas[i]) && (monotonic || decreaseAndCheckMaxUnavailable(scaleMaxUnavailable)) {
+		if isTerminating(replicas[i]) && monotonic {
 			klog.V(4).Infof(
 				"StatefulSet %s/%s is waiting for Pod %s to Terminate",
 				set.Namespace,
 				set.Name,
 				replicas[i].Name)
 			return &status, nil
+		} else if isTerminating(replicas[i]) && decreaseAndCheckMaxUnavailable(scaleMaxUnavailable) {
+			break
```
Comment (zmberg) [inline pkg/controller/statefulset/stateful_set_control.go:563]: klog.V(4).Infof(
			"StatefulSet %s/%s Pod %s is Terminating, and break pods scale",
			set.Namespace,
			set.Name,
			replicas[i].Name)

File: pkg/controller/statefulset/stateful_set_control.go [hunk 5]
```diff
@@ -591,6 +595,12 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 						replicas[i].Name)
 				}
 				return &status, nil
+			} else if !isAvailable && decreaseAndCheckMaxUnavailable(scaleMaxUnavailable) {
+				if waitTime > 0 {
+					// make sure we check later
+					durationStore.Push(getStatefulSetKey(set), waitTime)
+				}
+				break
```
Comment (zmberg) [inline pkg/controller/statefulset/stateful_set_control.go:603]: klog.V(4).Infof(
			"StatefulSet %s/%s Pod %s is unavailable, and break pods scale",
			set.Namespace,
			set.Name,
			replicas[i].Name)

**Comment by zmberg** [comment]
please *git commit -s*, and git push again.

**Comment by zmberg** [comment]
> please _git commit -s_, and git push again.

@ivanszl you need resolve the issue https://github.com/openkruise/kruise/pull/1136/checks?check_run_id=10079268512, otherwise there is no way to merge the pr.

**Pull request merged**